### PR TITLE
LF: get ride of user-defined patterns in ValueTranslation.

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
We prune the code of ValueTransalation from the user-defined patterns
in the code of Speedy's ValueTranslation. In particular we remove the usage of
com.daml.lf.language.Util.TTyConApp which:
- induces bad complexity,
- make difficult optimization.

CHANGELOG_BEGIN
CHANGELOG_END 

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala [hunk 1]
```diff
@@ -23,7 +22,7 @@ private[engine] final class ValueTranslator(compiledPackages: CompiledPackages)
   // this is not tail recursive, but it doesn't really matter, since types are bounded
   // by what's in the source, which should be short enough...
   @throws[PreprocessorException]
-  private def replaceParameters(params: ImmArray[(TypeVarName, Type)], typ0: Type): Type =
+  private[this] def replaceParameters(params: Traversable[(TypeVarName, Type)], typ0: Type): Type =
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:25]:
```suggestion
  private[this] def replaceParameters(params: Iterable[(TypeVarName, Type)], typ0: Type): Type =
```
scala 2.13 killed traversable

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala [hunk 2]
```diff
@@ -89,156 +96,175 @@ private[engine] final class ValueTranslator(compiledPackages: CompiledPackages)
...
-
-          // genMap
-          case (TGenMap(keyType, valueType), ValueGenMap(entries)) =>
-            SValue.SGenMap(
-              isTextMap = false,
-              entries = entries.iterator.map { case (k, v) =>
-                go(keyType, k, newNesting) -> go(valueType, v, newNesting)
-              },
-            )
+        def newNesting = nesting + 1
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:102]: why did you change this to a def instead of a val?

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala [hunk 3]
```diff
@@ -89,156 +95,184 @@ private[engine] final class ValueTranslator(compiledPackages: CompiledPackages)
...
+                  case (BTText, ValueText(t)) =>
+                    SValue.SText(t)
+                  case (BTParty, ValueParty(p)) =>
+                    SValue.SParty(p)
+                  case _ =>
+                    typeMismatch
+                }
+              case typeArg0 :: Nil =>
+                (bt, value) match {
+                  case (BTNumeric, ValueNumeric(d)) if typeArg0.isInstanceOf[TNat] =>
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:129]: I’m not that fond of `isInstanceOf`, can we pattern match instead?

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala [hunk 4]
```diff
@@ -89,156 +95,184 @@ private[engine] final class ValueTranslator(compiledPackages: CompiledPackages)
...
+                        entries = entries.iterator.map { case (k, v) =>
+                          SValue.SText(k) -> go(typeArg0, v, newNesting)
+                        },
+                      )
+                    }
+                  case _ =>
+                    typeMismatch
+                }
+              case typeArg0 :: typeArg1 :: Nil
+                  if bt == BTGenMap && value.isInstanceOf[ValueGenMap[_]] =>
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:165]: more `isinstanceOf` :disappointed:

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala [hunk 5]
```diff
@@ -89,156 +95,184 @@ private[engine] final class ValueTranslator(compiledPackages: CompiledPackages)
...
+                val ValueGenMap(entries) = value
+                if (entries.isEmpty) {
+                  SValue.SValue.EmptyTextMap
+                } else {
+                  SValue.SGenMap(
+                    isTextMap = true,
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:171]: I think this should be false.

**Review by cocreature — COMMENTED** [review]
Looks good but I’d like to get rid of the `asInstanceOf` before approving and merging.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala]: reverted

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala]: replaced.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala]: replaced as well.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala (hunk 5 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala]: good catch !. fixed.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala [hunk 6]
```diff
@@ -89,156 +97,190 @@ private[engine] final class ValueTranslator(compiledPackages: CompiledPackages)
...
+                    } else {
+                      SValue.SList(ls.map(go(typeArg0, _, newNesting)))
+                    }
+                  // textMap
+                  case (BTTextMap, ValueTextMap(entries)) =>
+                    if (entries.isEmpty) {
+                      SValue.SValue.EmptyTextMap
+                    } else {
+                      SValue.SGenMap(
+                        isTextMap = false,
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:160]: I think you replaced the wrong one? you are matching on `TextMap` here so this should be true. The one below where you’re matching on `GenMap` on the other hand should be false.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala [hunk 7]
```diff
@@ -89,156 +97,190 @@ private[engine] final class ValueTranslator(compiledPackages: CompiledPackages)
...
-                fail(
-                  s"Mismatching record id, the type tells us $typeRecordId, but the value tells us $id"
+                variantDef.constructorRank.get(constructorName) match {
+                  case None =>
+                    fail(
+                      s"Couldn't find provided variant constructor $constructorName in variant $tyCon"
+                    )
+                  case Some(rank) =>
+                    val (_, argTyp) = variantDef.variants(rank)
+                    val replacedTyp =
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:209]: is `toSeq.view` actually more efficient than just `map`?

**Review by cocreature — APPROVED** [review]
LGTM apart from the `isTextMap` fields which I think are now both wrong.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala (hunk 6 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:160]: I need another coffee.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala (hunk 7 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:209]: In this case, I think it is.

First I need the result to be `Iterable` (to be able to use `.toMap`). 

1 - Since `dataTypParams` is an ImmArray I need a `.toSeq` somewhere. 
2 - `dataTypParams.toSeq.map(_._1).zip(tyArgs)` will allocate 2 intermediate `ImmArraySeq`
3 - In general Array builder allocate several intermediate arrays (as they do not know the final size of the array they are building).

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala (hunk 6 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala]: I had a test.

Comment (cocreature) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala]: who needs coffee if you have tests instead :slightly_smiling_face:
